chips/nrf52/src/usbd.rs

```diff
@@ -0,0 +1,562 @@
+    /// - Address: 0x49C - 0x4A0
+    wlengthh: ReadOnly<u32, Byte::Register>,
+    /// Amount of bytes received last in the data stage of
+    /// this OUT\[n\] endpoint
+    /// - Address: 0x4A0 - 0x4C0
+    size_epout: [ReadWrite<u32, EndpointSize::Register>; NUM_ENDPOINTS],
+    /// Amount of bytes received last on this iso OUT data endpoint
+    /// - Address: 0x4C0 - 0x4C4
+    size_iosout: ReadOnly<u32, IsoEndpointSize::Register>,
+    _reserved10: [u32; 15],
+    /// Enable USB
+    /// - Address: 0x500 - 0x504
+    enable: ReadWrite<u32, Usb::Register>,
+    /// Control of the USB pull-up
+    /// - Address: 0x504 - 0x508
+    usbpullup: ReadWrite<u32, UsbPullup::Register>,
+    /// State at which the DPDMDRIVE task will force D+ and D-.
+    /// The DPDMNODRIVE task reverts the control of the lines
+    /// to MAC IP (no forcing).
+    /// - Address: 0x508 - 0x50C
+    dpdmvalue: ReadWrite<u32, DpDmValue::Register>,
+    /// Data toggle control and status
+    /// - Address: 0x50C - 0x510
+    dtoggle: ReadWrite<u32, Toggle::Register>,
+    /// Endpoint IN enable
+    /// - Address: 0x510 - 0x514
+    epinen: ReadWrite<u32, EndpointEnable::Register>,
+    /// Endpoint OUT enable
+    /// - Address: 0x514 - 0x518
+    epouten: ReadWrite<u32, EndpointEnable::Register>,
+    /// STALL endpoints
+    /// - Address: 0x518 - 0x51C
+    epstall: WriteOnly<u32, EndpointStall::Register>,
+    /// Controls the split of ISO buffers
+    /// - Address: 0x51C - 0x520
+    isosplit: ReadWrite<u32, IsoSplit::Register>,
+    /// Returns the current value of the start of frame counter
+    /// - Address: 0x520 - 0x524
+    framecntr: ReadOnly<u32, FrameCounter::Register>,
+    _reserved11: [u32; 2],
+    /// Controls USBD peripheral low power mode during USB suspend
+    /// - Address: 0x52C - 0x530
+    lowpower: ReadWrite<u32, LowPower::Register>,
+    /// Controls the response of the ISO IN endpoint to an IN token
+    /// when no data is ready to be sent
+    /// - Address: 0x530 - 0x534
+    isoinconfig: ReadWrite<u32, IsoInConfig::Register>,
+    _reserved12: [u32; 51],
+    /// - Address: 0x600 - 0x6A0
+    epin: [EndpointRegisters; NUM_ENDPOINTS],
+    /// - Address: 0x6A0 - 0x6B4
+    isoin: EndpointRegisters,
+    _reserved13: [u32; 19],
+    /// - Address: 0x700 - 0x7A0
+    epout: [EndpointRegisters; NUM_ENDPOINTS],
+    /// - Address: 0x7A0 - 0x7B4
+    isoout: EndpointRegisters,
+    _reserved14: [u32; 19],
+    /// Errata 166 related register (ISO double buffering not functional)
+    /// - Address: 0x800 - 0x804
+    errata166_1: WriteOnly<u32>,
+    /// Errata 166 related register (ISO double buffering not functional)
+    /// - Address: 0x804 - 0x808
+    errata166_2: WriteOnly<u32>,
+    _reserved15: [u32; 261],
+    /// Errata 199 related register (USBD cannot receive tasks during DMA)
+    /// - Address: 0xC1C - 0xC20
+    errata199: WriteOnly<u32>,
+}
+
+#[repr(C)]
+struct EndpointRegisters {
+    ptr: VolatileCell<*const u8>,
```

Comment:
I think the intention here is to have a type that encapsulates the translation of a `&'static [u8]` to raw hardware registers. That's very good, since we'd like to trust as little of even this module to perform that translation correctly. The EasyDMA is probably the most sensitive part of the chip in terms of isolation.

On the incoming end (is that IN or OUT? I can never remember USB's perspective), this is great since reading a `*const u8` requires unsafe, so there is no danger in exposing this interface.

For the other side, it could maybe use _slightly_ more protection. Anything can create an arbitrary `*const u8` (it's a simple "safe" cast from a `usize`), so anyone who can access these fields could basically "leak" arbitrary memory from anywhere on the chip. Clearly these fields are private to this module, so the "attacker" (hopefully at worst buggy code) is limited to these few hundred lines.

However, I'd feel more comfortable if these fields were hidden in a submodule and had constructors and accessor methods for creating an `EndpointRegisters` and retrieving from one something like a `&'static [u8]`.

Reply:
Note that we haven't done a good job of this in the nRF port in general (i.e. the UART, SPI, etc), but this is maybe as good a place to start as any (especially since USB is likely to be a much more complex driver overall).

Now that I look, I realize we've lost this in the SAM4L as well since at least the registers interface, which isn't great...